lib/Dialect/TritonNvidiaGPU/IR/Dialect.cpp

```diff
@@ -51,51 +51,38 @@ namespace nvidia_gpu {
 static constexpr int numTmemRows = 128;
 
 TMemAllocation getTmemAllocSizes(MemDescType memDescType) {
-  const int rowSizeInBytes = 4;
-  auto shapePerCTA = triton::gpu::getShapePerCTA(memDescType);
-  if (isa<TensorMemoryScalesEncodingAttr>(memDescType.getEncoding())) {
-    // For scales the data are packed and replicated 4 times.
-    assert(memDescType.getElementType().getIntOrFloatBitWidth() == 8);
-    assert(memDescType.getShape().size() == 2 &&
-           "TODO handle multibuffering of scales.");
-    int k = shapePerCTA[1];
-    int m = shapePerCTA[0];
-    int numColumn = 2 * ceil<int>(m, 64) * ceil<int>(k, 4);
-    return TMemAllocation(numColumn, numTmemRows);
+  auto *ctx = memDescType.getContext();
+  auto S = [&](StringRef str) { return StringAttr::get(ctx, str); };
+  auto kRow = S("row");
+  auto kCol = S("col");
+  // Remove multibuffering if present
+  auto shape = memDescType.getShape().take_back(2);
+  auto ll = toLinearLayout(shape, memDescType.getEncoding());
+  auto bitwidth = memDescType.getElementTypeBitWidth();
+  int nRow = ll.getInDimSize(kRow);
+  int nCol = ll.getInDimSize(kCol) / (32 / bitwidth);
+  // If we have just one 16xcol block per warp, we don't allocate 128 rows
+  // we use 64 rows instead.
+  // We could generalise this to when we have more zeros in the layout, but
+  // the allocator does not support this yet
+  if (ll.getBasis(kRow, llvm::Log2_32(16)) == ArrayRef{0, 0}) {
+    nRow /= 2;
   }
-  assert(isa<triton::nvidia_gpu::TensorMemoryEncodingAttr>(
-             memDescType.getEncoding()) &&
-         "Expecting a tensor memory encoding attribute");
-  triton::nvidia_gpu::TensorMemoryEncodingAttr attr =
-      cast<triton::nvidia_gpu::TensorMemoryEncodingAttr>(
-          memDescType.getEncoding());
-  unsigned bitwidth = memDescType.getElementType().getIntOrFloatBitWidth();
-  int64_t elementSizeInBytes = (bitwidth * attr.getColStride()) / 8;
-  int sizeInBytes = product(shapePerCTA) * elementSizeInBytes;
-  int numRows = numTmemRows;
-  // BlockM of 64 is and interleaved format, where for single message only the
```

Comment:
nit: can we keep this drawing somewhere?

Reply:
I was going to move it to NvidiaGPUAttrDefs.td but I found there is already a similar drawing there:
https://github.com/triton-lang/triton/blob/382e7c7be4c3d9bf48370c93393125102181205e/include/triton/Dialect/TritonNvidiaGPU/IR/TritonNvidiaGPUAttrDefs.td#L18-L30
do you still want me to move it somewhere else?